### PR TITLE
Add duration formatter

### PR DESCRIPTION
**Pull request by bim9262**
TODO: document duration formatter.

Any thoughts on the names of the arguments, the format, or the defaults?

The formatted can be used by the battery, uptime, tea_timer and pomodoro blocks.

### CODE REVIEW

File: src/formatting/formatter/duration.rs
Comment (bim9262) [inline src/formatting/formatter/duration.rs:65]: This is borrowed from the eng formatter, but the units here aren't prefixes

**Comment by MaxVerevkin** [comment]
It is hard to judge/review until there is at least one block using this. The good thing is that blocks can be migrated in a non-breaking way, for example `tea_timer` can still have the old (and eventually deprecated and then removed) hours/minutes/seconds, in addition to the new placeholder.

By the way, I think it is safe to delete `fix.rs` for now.

**Comment by bim9262** [comment]
Added two examples, uptime and tea_timer. Still need to doc the formatter...

File: src/blocks/tea_timer.rs
Comment (MaxVerevkin) [inline src/blocks/tea_timer.rs:91]: The state change seems unrelated

**Comment by MaxVerevkin** [comment]
Do I understand correctly that currently there is no way to use `mm:ss` format? I think this can be achieved by allowing to replace the space character between units with `:` and adding an option to hide units. Maybe we can even add a shortcut parameter to set all these values, because `.dur(units:6,show_leading_units_if_zero:false,separator:':',show_units:false)` is too much.

By the way, maybe `show_leading_units_if_zero` should default to `false`? I'm not %100 sure though.

File: src/formatting/formatter/duration.rs
Comment (MaxVerevkin) [inline src/formatting/formatter/duration.rs:163]:
```suggestion
                                        != self.min_unit_index + 1 - self.units)
```

because

```toml
[[block]]
block = "tea_timer"
format = " $time.dur(units:6) | click to start "
```

causes a panic:

```
thread 'main' panicked at src/formatting/formatter/duration.rs:163:44:
attempt to subtract with overflow
```

**Comment by bim9262** [comment]
> Do I understand correctly that currently there is no way to use `mm:ss` format? I think this can be achieved by allowing to replace the space character between units with `:` and adding an option to hide units. Maybe we can even add a shortcut parameter to set all these values, because `.dur(units:6,show_leading_units_if_zero:false,separator:':',show_units:false)` is too much.

That's right. If you wanted to do [hh]:mm:ss you'd want to do something like `.dur(max_unit:h,min_unit:s,show_leading_units_if_zero:false,separator:':',show_units:false)` otherwise after you get to a day+ you'd end up with a format like `dd:hh:mm:ss`, which probably isn't a good format.

The most flexible solution would be to allow an explicit format to be specified if we want to allow "sub-formats" like:
```.dur(format:'{$d |}{$h:|}{$m:|}'{$s:|}',show_leading_units_if_zero:false)```

or using the allow argument mentioned in https://github.com/greshake/i3status-rust/issues/1957:
```.dur(format:'{$d.eng(allow:1..) |}{$h.eng(allow:1..):|}{$m.eng(allow:1..):|}'{$s.eng(allow:1..):|}')```

> By the way, maybe `show_leading_units_if_zero` should default to `false`? I'm not %100 sure though.

I set `show_leading_units_if_zero` to true, because it means that say you have a countdown timer you you would go from `" 1m  0s"` to `" 0m 59s"` to keep a consistent width.

File: src/blocks/tea_timer.rs
Comment (bim9262) [reply src/blocks/tea_timer.rs:91]: That's true. I was mirroring something I did in the pomodoro block (which also needed this formatter).
I can remove it from this PR if you'd like.

**Comment by bim9262** [comment]
Another option would be to  use [strftime](https://docs.rs/chrono/latest/chrono/format/strftime/index.html) style formats.

**Comment by bim9262** [comment]
Perhaps adding an `hms` option which sets the max_arg to hours and sets the separator to : and hides the units would be enough instead of adding nested formatters or a new parser.

**Comment by MaxVerevkin** [comment]
> The most flexible solution would be to allow an explicit format to be specified if we want to allow "sub-formats" like:
> `.dur(format:'{$d |}{$h:|}{$m:|}'{$s:|}',show_leading_units_if_zero:false)`

That's... too much formatting :smile: 

> That's right. If you wanted to do [hh]:mm:ss you'd want to do something like `.dur(max_unit:h,min_unit:s,show_leading_units_if_zero:false,separator:':',show_units:false)` otherwise after you get to a day+ you'd end up with a format like `dd:hh:mm:ss`, which probably isn't a good format.
>
> ...
>
> Perhaps adding an `hms` option which sets the max_arg to hours and sets the separator to : and hides the units would be enough instead of adding nested formatters or a new parser.

`.dur(hms:true)` seems pretty good.

> or using the allow argument mentioned in #1957:
> `.dur(format:'{$d.eng(allow:1..) |}{$h.eng(allow:1..):|}{$m.eng(allow:1..):|}'{$s.eng(allow:1..):|}')`

Again, nested format strings feel weird. But maybe... maybe we can add support for some sort of "fields", like `$time.h.eng(w:2):$time.s.eng(w:2)`? Not sure if it makes sense.

**Comment by bim9262** [comment]
> > Perhaps adding an `hms` option which sets the max_arg to hours and sets the separator to : and hides the units would be enough instead of adding nested formatters or a new parser.
> 
> `.dur(hms:true)` seems pretty good.
> 
> > or using the allow argument mentioned in #1957:
> > `.dur(format:'{$d.eng(allow:1..) |}{$h.eng(allow:1..):|}{$m.eng(allow:1..):|}'{$s.eng(allow:1..):|}')`
> 
> Again, nested format strings feel weird. But maybe... maybe we can add support for some sort of "fields", like `$time.h.eng(w:2):$time.s.eng(w:2)`? Not sure if it makes sense.

I'm not sure that the field concept would be generally useful (outside of the duration format) and much more verbose so I'm leaning towards the `hms` option.

**Comment by bim9262** [comment]
I think the behaviour of `round_up`  should be modified from `Round up to the nearest min_unit`  to `Round up to the nearest minimum displayed unit`

**Comment by bim9262** [comment]
I guess we should decide if hms should be enabled by default for blocks that used that format previously or if it is better to use the unambiguous default format. (2 units of hms are either hh:mm or mm::ss, but unless you know the interval/or it's a small interval and you can see the seconds changing)

**Comment by bim9262** [comment]
Once we think the settings/defaults make sense I'll add some unit tests as well.

**Comment by bim9262** [comment]
@MaxVerevkin I know this is a lot to review, do you want me to move the splitting up of the formatters into modules into its own PR?

**Comment by MaxVerevkin** [comment]
> do you want me to move the splitting up of the formatters into modules into its own PR?

That would be nice :)

**Comment by bim9262** [comment]
@MaxVerevkin, I rebased on master now that the formatter split and allow empty strings PRs have merged.
Any changes to the argument names or default values?
Should we make the default of the tea timer block use hms by default, since that's the default format now?

**Comment by MaxVerevkin** [comment]
> Should we make the default of the tea timer block use hms by default, since that's the default format now?

Yes, let's try to keep the defaults as they are.

I'll try to take a closer look tomorrow.

File: src/blocks/tea_timer.rs
Comment (MaxVerevkin) [reply src/blocks/tea_timer.rs:93]: Let's remove this for now, maybe add it later in a new PR.

**Comment by MaxVerevkin** [comment]
The [types](https://greshake.github.io/i3status-rust/i3status_rs/formatting/index.html#types) section of the docs needs to be updated.

File: src/formatting.rs
Comment (MaxVerevkin) [inline src/formatting.rs:83]: I think it would be very useful to have a list of all possible units. For example below the table, with a note "see below for the list of all possible units".

**Review by MaxVerevkin — APPROVED** [review]
Looks good :+1:

**Comment by MaxVerevkin** [comment]
@bim9262 Is this ready to be merged?

**Comment by bim9262** [comment]
I still wanted to add some tests

**Comment by bim9262** [comment]
@MaxVerevkin I've added some tests :)

File: src/formatting/formatter/duration.rs
Comment (MaxVerevkin) [inline src/formatting/formatter/duration.rs:266]: `eng.rs` has a similar macro for tests, let's share them?

**Review by MaxVerevkin — COMMENTED** [review]
This currently changes the way `$time` is rendered in battery block, because `hms` is not the default and the behavior is preserved only if explicit `.str()` is used. I think there are two things we can do:

- Make `Value::Duration` hold the default value for `hms`.
- Do not touch `time` and instead deprecate it and add `remaining`/`duration`/something.

**Comment by bim9262** [comment]
> * Make `Value::Duration` hold the default value for `hms`.

Can you explain what you mean? Hold the string value like I did with `time`

> * Do not touch `time` and instead deprecate it and add `remaining`/`duration`/something.

Got it, I can add another value instead changing the type of time.

File: src/formatting/formatter/duration.rs
Comment (bim9262) [reply src/formatting/formatter/duration.rs:266]: Created #2054, will rebase after that's merged

**Comment by MaxVerevkin** [comment]
> > * Make `Value::Duration` hold the default value for `hms`.
> 
> Can you explain what you mean? Hold the string value like I did with `time`

Something like `Value::duration(dur).with_default_hms()`, `DurationFormatter::hms` would need to be changed from `bool` to `Option<bool>`. Not sure if this is a good idea though.

**Comment by bim9262** [comment]
> Something like `Value::duration(dur).with_default_hms()`, 

The format for `time` in the battery block isn't even the default hms format, it doesn't display seconds, just hh:mm.

>`DurationFormatter::hms` would need to be changed from `bool` to `Option<bool>`. Not sure if this is a good idea though.

I had thought about doing something like:

```rs
#[derive(Debug, Default)]
pub struct DurationFormatter(Option<DurationFormatterInner>);

#[derive(Debug, Default)]
pub struct DurationFormatterInner {
    hms: bool,
    max_unit_index: usize,
    min_unit_index: usize,
    units: usize,
    round_up: bool,
    unit_has_space: bool,
    pad_with: PadWith,
    show_leading_units_if_zero: bool,
}
 ```
but I dislike having different defaults for different blocks.

So I think I'll follow the approach: 

>Do not touch time and instead deprecate it and add remaining/duration/something.

**Comment by bim9262** [comment]
@MaxVerevkin anything else needed here or are we good to merge?

File: src/formatting.rs
Comment (MaxVerevkin) [inline src/formatting.rs:89]:
```suggestion
//! `unit_space`                 | Should there be a space between the value and unit symbol (not allowed when `hms:true`)          | `false`
```

File: src/formatting.rs
Comment (MaxVerevkin) [inline src/formatting.rs:91]: I think something shorter would be nice :) Maybe `leading_zeroes`?

Comment (MaxVerevkin) [reply src/formatting.rs:91]: I think you forgot to update the docs.

**Comment by bim9262** [comment]
@MaxVerevkin, I made the suggested changes

File: src/formatting.rs
Comment (bim9262) [reply src/formatting.rs:91]: Ah, because I forgot to hit save before commiting

**Review by MaxVerevkin — APPROVED** [review]
LGTM! Again, sorry for late review.

**Comment by bim9262** [comment]
> LGTM! Again, sorry for late review.

No worries, I've also been late on the reviews too...

**Pull request merged**
